File: 28-find-the-index-of-the-first-occurrence-in-a-string/find-the-index-of-the-first-occurrence-in-a-string.py
class Solution:
    def strStr(self, haystack: str, needle: str) -> int:
        hay_len = len(haystack)
        needle_len = len(needle)

        if needle_len > hay_len:
            return -1

        for index in range(hay_len - needle_len + 1):
            if haystack[index: index + needle_len] == needle:
                return index
        return -1            

File: 28-find-the-index-of-the-first-occurrence-in-a-string/find-the-index-of-the-first-occurrence-in-a-string.java
class Solution {
    public int strStr(String haystack, String needle) {
        // if needle is empty
        if (needle == null || needle.length() == 0) {
            return 0;
        }
        // if needle is longer than haystack
        if (needle.length() > haystack.length()) {
            return -1;
        }

        // bounds check. haystack can't be shorter than needle
        int bounds = haystack.length() - needle.length();
        
        for (int i = 0; i <= bounds; i++) {
            if (haystack.charAt(i) != needle.charAt(0)) {
                continue;
            }    
            int j;
            for (j = 1; j < needle.length(); j++) {
                if (haystack.charAt(i + j) != needle.charAt(j)) {
                    break;
                }
            }

            // if we matched the whole needle
            if (j == needle.length()) {
                return i;
            }
        }
        return -1;

    }
}